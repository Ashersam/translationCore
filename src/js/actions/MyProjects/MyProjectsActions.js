import consts from '../ActionTypes';
// helpers
import * as myProjectsHelpers from '../../helpers/myProjectsHelpers';
import { getTranslate } from '../../selectors';
import { confirmAction } from '../../middleware/confirmation/confirmationMiddleware';

/**
 * @description With the list of project directories, generates an array of project detail objects
 */
export function getMyProjects() {
  return ((dispatch, getState) => {
    myProjectsHelpers.migrateResourcesFolder();
    const state = getState();
    const { projectDetailsReducer: { projectSaveLocation } } = state;
    let projects = myProjectsHelpers.getProjectsFromFS(projectSaveLocation, null);

    dispatch({
      type: consts.GET_MY_PROJECTS,
      projects: projects,
    });
  });
}

/**
 * Moves a project into the archive after the user confirms.
 * Archived projects can be restored at a later time.
 * @param projectPath {string} the path to the project that will be archived.
 */
export const archiveProject = (projectPath) => (dispatch, getState) => {
  const translate = getTranslate(getState());

  dispatch(confirmAction({
    message: translate('projects.confirm_archive'),
    confirmButtonText: translate('projects.archive_project'),
  }, (disp) => {
    // TODO: archive the project
    disp({
      type: consts.ARCHIVE_PROJECT,
      path: projectPath,
    });
  }));
};
