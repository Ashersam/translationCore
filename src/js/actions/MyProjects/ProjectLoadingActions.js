import consts from '../ActionTypes';
import path from 'path-extra';
import ospath from 'ospath';
// actions
import migrateProject from '../../helpers/ProjectMigration';
import {initializeReducersForProjectOpenValidation, validateProject} from '../Import/ProjectValidationActions';
import * as BodyUIActions from '../BodyUIActions';
import * as RecentProjectsActions from '../RecentProjectsActions';
import {openAlertDialog} from '../AlertModalActions';
import * as ProjectDetailsActions from '../ProjectDetailsActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
//helpers
import * as manifestHelpers from '../../helpers/manifestHelpers';
import {
  getActiveLocaleLanguage,
  getProjectManifest, getSourceBook, getTargetBook, getToolGatewayLanguage,
  getTools,
  getTranslate,
  getUsername
} from "../../selectors";
import {isProjectSupported} from '../../helpers/ProjectValidation/ProjectStructureValidationHelpers';
import { loadBookTranslations } from "../ResourcesActions";
import ProjectAPI from "../../helpers/ProjectAPI";
import CoreAPI from "../../helpers/CoreAPI";
import { copyGroupDataToProject } from "../../helpers/ResourcesHelpers";

// constants
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

function delay(ms) {
  return new Promise ((resolve) =>
    setTimeout(resolve, ms)
  );
}

/**
 * This thunk opens a project and prepares it for use in tools.
 * @param {string} name -  the name of the project
 */
export const openProject = (name) => {
  return async (dispatch, getState) => {
    const projectDir = path.join(PROJECTS_PATH, name);
    const translate = getTranslate(getState());

    try {
      dispatch(resetReducersData());
      dispatch(initializeReducersForProjectOpenValidation());
      dispatch(
        openAlertDialog(translate('projects.loading_project_alert'), true));
      // TRICKY: prevent dialog from flashing on small projects
      await delay(200);
      await isProjectSupported(projectDir, translate);
      migrateProject(projectDir, null, getUsername(getState()));
      await dispatch(validateProject(projectDir));

      // load the book data
      const manifest = getProjectManifest(getState());
      await dispatch(loadBookTranslations(manifest.project.id, name));

      // connect the tools
      const tools = getTools(getState());
      for (const t of tools) {
        // copy group data
        const language = getToolGatewayLanguage(getState(), t.name);
        copyGroupDataToProject(language, t.name, projectDir);

        // connect tool api
        const toolProps = makeToolProps(dispatch, getState(), projectDir, manifest.project.id);
        t.api.triggerWillConnect(toolProps);
      }

      await dispatch(displayTools());
    } catch (e) {
      // TODO: clean this up
      if (e.type !== 'div') console.warn(e);
      // clear last project must be called before any other action.
      // to avoid triggering autosaving.
      dispatch(clearLastProject());
      dispatch(openAlertDialog(e));
      dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
    }
  };
};

function resetReducersData() {
  // TODO: this is crazy. All of related reducers could be keyed by the same action.
  return (dispatch => {
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_DATA });
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_INDEX });
    dispatch({ type: consts.CLEAR_CONTEXT_ID });
    dispatch({ type: consts.CLEAR_ALIGNMENT_DATA });
    dispatch({ type: consts.CLEAR_RESOURCES_REDUCER });
    dispatch({ type: consts.CLEAR_PREVIOUS_FILTERS});
  });
}

/**
 * TODO: this is very similar to what is in the {@link ToolContainer} and probably needs to be abstracted.
 * This is just a temporary prop generator until we can properly abstract the tc api.
 * @param dispatch
 * @param state
 * @param projectDir
 * @param bookId
 * @returns {*}
 */
function makeToolProps(dispatch, state, projectDir, bookId) {
  const projectApi = new ProjectAPI(projectDir);
  const coreApi = new CoreAPI(dispatch);
  const {code} = getActiveLocaleLanguage(state);
  const sourceBook = getSourceBook(state);
  const targetBook = getTargetBook(state);

  return {
    // project api
    readProjectDir: projectApi.readDataDir,
    readProjectDirSync: projectApi.readDataDirSync,
    writeProjectData: projectApi.writeDataFile,
    writeProjectDataSync: projectApi.writeDataFileSync,
    readProjectData: projectApi.readDataFile,
    readProjectDataSync: projectApi.readDataFileSync,
    projectDataPathExists: projectApi.dataPathExists,
    projectDataPathExistsSync: projectApi.dataPathExistsSync,
    deleteProjectFile: projectApi.deleteDataFile,

    // tC api
    showDialog: coreApi.showDialog,
    showLoading: coreApi.showLoading,
    closeLoading: coreApi.closeLoading,
    showIgnorableAlert: coreApi.showIgnorableAlert,
    appLanguage: code,

    // project data
    sourceBook,
    targetBook,

    contextId: {
      reference: {
        bookId,
        chapter: "1", // TRICKY: just some dummy values at first
        verse: "1"
      }
    },

    // deprecated props
    showIgnorableDialog: (...args) => {
      console.warn('DEPRECATED: showIgnorableDialog is deprecated. Use showIgnorableAlert instead');
      return coreApi.showIgnorableAlert(...args);
    },
    get toolsReducer () {
      console.warn(`DEPRECATED: toolsReducer is deprecated.`);
      return {};
    },
    projectFileExistsSync: (...args) => {
      console.warn(`DEPRECATED: projectFileExistsSync is deprecated. Use pathExistsSync instead.`);
      return projectApi.dataPathExistsSync(...args);
    },
    get targetBible() {
      console.warn('DEPRECATED: targetBible is deprecated. Use targetBook instead');
      return targetBook;
    },
    get sourceBible() {
      console.warn('DEPRECATED: sourceBible is deprecated. Use sourceBook instead');
      return sourceBook;
    },
  };
}

/**
 * @description - Opening the tools screen upon making sure the project is
 * not a titus or in the user is in developer
 */
export function displayTools() {
  return (dispatch, getState) => {
    const state = getState();
    const translate = getTranslate(state);
    return new Promise ((resolve, reject) => {
      try {
        const { currentSettings } = state.settingsReducer;
        const { manifest } = state.projectDetailsReducer;
        if (manifestHelpers.checkIfValidBetaProject(manifest) || currentSettings.developerMode) {
          // Go to toolsCards page
          dispatch(BodyUIActions.goToStep(3));
        } else {
          dispatch(RecentProjectsActions.getProjectsFromFolder());
          reject(translate('projects.books_available', {app: translate('_.app_name')}));
        }
      } catch (error) {
        console.error(error);
        reject(error);
      }
      resolve();
    });
  };
}

/**
 * @description - Wrapper to clear everything in the store that could
 * prevent a new project from loading
 */
export function clearLastProject() {
  return (dispatch) => {
    /**
     * ATTENTION: THE project details reducer must be reset
     * before any other action being called to avoid
     * autosaving messing up with the project data.
     */
    dispatch({ type: consts.RESET_PROJECT_DETAIL });
    dispatch(BodyUIActions.toggleHomeView(true));
    dispatch(ProjectDetailsActions.resetProjectDetail());
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_DATA });
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_INDEX });
    dispatch({ type: consts.CLEAR_CONTEXT_ID });
    dispatch({ type: consts.CLOSE_TOOL });
    dispatch({ type: consts.CLEAR_RESOURCES_REDUCER });
    dispatch({ type: consts.CLEAR_PREVIOUS_FILTERS});
  };
}

/**
 * @description loads and set the projects details into the projectDetailsReducer.
 * @param {string} projectPath - path location in the filesystem for the project.
 * @param {object} manifest - project manifest.
 */
export function loadProjectDetails(projectPath, manifest) {
  return (dispatch) => {
    dispatch(ProjectDetailsActions.setSaveLocation(projectPath));
    dispatch(ProjectDetailsActions.setProjectManifest(manifest));
  };
}
