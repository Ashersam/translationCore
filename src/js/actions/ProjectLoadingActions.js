import path from 'path-extra';
import consts from './ActionTypes';
// actions
import * as TargetLanguageActions from './TargetLanguageActions';
import * as AlertModalActions from './AlertModalActions';
// helpers
import * as ProjectLoadingHelpers from '../helpers/ProjectLoadingHelpers';

/**
 * @description function that handles both getGroupsIndex and
 * getGroupData with promises.
 * @param {string} toolName - name of the tool being loaded.
 * @return {object} object action.
 */
export function loadProjectData(toolName) {
  return ((dispatch, getState) => {
    return new Promise((resolve, reject) => {
      let { projectDetailsReducer } = getState();
      let { projectSaveLocation, params } = projectDetailsReducer;
      const dataDirectory = path.join(projectSaveLocation, '.apps', 'translationCore', 'index', toolName);

      dispatch(TargetLanguageActions.generateTargetBible(projectSaveLocation));
      ProjectLoadingHelpers.getGroupsIndex(dispatch, toolName, dataDirectory)
        .then(() => {
          ProjectLoadingHelpers.getGroupData(dispatch, dataDirectory, toolName, params)
          .then(dispatch({ type: consts.TOGGLE_LOADER_MODAL, show: false }))
        });
    })
    .catch(err => {
      console.warn(err);
      AlertModalActions.openAlertDialog("Oops! We have encountered a problem loading your project. Please contact Help Desk (help@door43.org) for assistance.");
    });
  })
}
