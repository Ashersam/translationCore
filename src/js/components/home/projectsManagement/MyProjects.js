import React from 'react';
import PropTypes from 'prop-types';
// components
import ProjectCard from './ProjectCard';

let MyProjects = ({myProjects, user, actions, translate}) => {
  let projects = myProjects.map((projectDetails, index) =>
    <ProjectCard user={user}
                 key={index}
                 translate={translate}
                 projectDetails={projectDetails}
                 actions={actions} />
  );

  if(myProjects.length == 0) {
    projects.push(
      <p key={0}><br/>
        <b>
          {translate('home.project.no_projects')}
        </b>
      </p>
    );
  }

  return (
    <div style={{ height: '100%' }}>
      Projects
      <div style={{ height: '100%', overflowY: 'auto', paddingRight: '10px' }}>
        {projects}
      </div>
    </div>
  );
};

MyProjects.propTypes = {
  translate: PropTypes.func.isRequired,
  myProjects: PropTypes.array.isRequired,
  user: PropTypes.any,
  actions: PropTypes.object.isRequired
};

export default MyProjects;
