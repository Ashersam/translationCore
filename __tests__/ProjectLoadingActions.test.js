import configureMockStore from 'redux-mock-store';
import fs from 'fs-extra';
import thunk from 'redux-thunk';
import path from 'path-extra';
import ospath from 'ospath';
import * as ProjectLoadingActions from "../src/js/actions/MyProjects/ProjectLoadingActions";
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const importProjectName = 'project';
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

jest.mock('fs-extra');

jest.mock('../src/js/selectors', () => ({
  getActiveLocaleLanguage: () => {
    return {code: 'en'};
  },
  getTranslate: () => {
    return jest.fn((code) => {
      return code;
    });
  }
}));


describe('ProjectLoadingActions.migrateValidateLoadProject', () => {
  let initialState = {};
  const projectName = 'en_tit';
  const sourcePath = "__tests__/fixtures/project/";
  const projectPath = path.join(PROJECTS_PATH,projectName);
  beforeEach(() => {
    // reset mock filesystem data
    fs.__resetMockFS();
    let copyFiles = [projectName];
    fs.__loadFilesIntoMockFs(copyFiles, sourcePath, PROJECTS_PATH);

    initialState = {
      homeScreenReducer: {
        stepper: {
          stepIndex: 1,
          nextStepName: 'Project Information',
          previousStepName: 'Cancel',
          nextDisabled: false
        }
      },
      loginReducer: {
        loggedInUser: false,
        userdata: {},
        feedback: '',
        subject: 'Bug Report',
        placeholder: 'Leave us your feedback!'
      },
      projectDetailsReducer: {
        projectSaveLocation: '',
        manifest: {},
        currentProjectToolsProgress: {},
        projectType: null
      },
      localImportReducer: {
        selectedProjectFilename: '',
        sourceProjectPath: '',
        oldSelectedProjectFileName: null
      },
      projectInformationCheckReducer: {}
    };
  });

  it('should open valid project', () => {
    // given
    const store = mockStore(initialState);
    const selectedProjectFilename = projectName;
    // when
    return store.dispatch(ProjectLoadingActions.migrateValidateLoadProject(selectedProjectFilename)).then(() => {
      //then
      expect(store.getActions()).toMatchSnapshot();
    });
  }, 5000);

  it('should not open invalid project', () => {
    // given
    const store = mockStore(initialState);
    const selectedProjectFilename = 'test';
    // when
    return store.dispatch(ProjectLoadingActions.migrateValidateLoadProject(selectedProjectFilename)).then(() => {
      //then
      expect(store.getActions()).toMatchSnapshot();
    });
  }, 5000);

});
