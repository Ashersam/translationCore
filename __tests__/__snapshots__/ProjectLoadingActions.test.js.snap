// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectLoadingActions.migrateValidateLoadProject should not open invalid project 1`] = `
Array [
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "alreadyImported": true,
    "type": "SET_ALREADY_IMPORTED_IN_PROJECT_INFORMATION_CHECK_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "boolean": true,
    "type": "TOGGLE_HOME_VIEW",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_DATA",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_INDEX",
  },
  Object {
    "type": "CLEAR_CONTEXT_ID",
  },
  Object {
    "type": "CLEAR_CURRENT_TOOL_DATA",
  },
  Object {
    "type": "CLEAR_RESOURCES_REDUCER",
  },
  Object {
    "type": "CLEAR_PREVIOUS_FILTERS",
  },
  Object {
    "currentToolTitle": "",
    "type": "SET_CURRENT_TOOL_TITLE",
  },
  Object {
    "alertMessage": [TypeError: Cannot read property 'bookId' of undefined],
    "loading": undefined,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "showProjectValidationStepper": false,
    "type": "TOGGLE_PROJECT_VALIDATION_STEPPER",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "boolean": true,
    "type": "TOGGLE_HOME_VIEW",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_DATA",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_INDEX",
  },
  Object {
    "type": "CLEAR_CONTEXT_ID",
  },
  Object {
    "type": "CLEAR_CURRENT_TOOL_DATA",
  },
  Object {
    "type": "CLEAR_RESOURCES_REDUCER",
  },
  Object {
    "type": "CLEAR_PREVIOUS_FILTERS",
  },
  Object {
    "currentToolTitle": "",
    "type": "SET_CURRENT_TOOL_TITLE",
  },
  Object {
    "type": "CLEAR_COPYRIGHT_CHECK_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "type": "CLEAR_MERGE_CONFLICTS_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "projects": Array [
      Object {
        "accessTimeAgo": "in a few seconds",
        "bookAbbr": "tit",
        "bookName": "Titus",
        "isSelected": false,
        "projectName": "en_tit",
        "projectSaveLocation": "/Users/blm/translationCore/projects/en_tit",
        "target_language": Object {
          "direction": "ltr",
          "id": "en",
          "name": "English",
        },
      },
    ],
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "LOADED_ONLINE_FAILED",
  },
]
`;

exports[`ProjectLoadingActions.migrateValidateLoadProject should open valid project 1`] = `
Array [
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "alreadyImported": true,
    "type": "SET_ALREADY_IMPORTED_IN_PROJECT_INFORMATION_CHECK_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "boolean": true,
    "type": "TOGGLE_HOME_VIEW",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_DATA",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_INDEX",
  },
  Object {
    "type": "CLEAR_CONTEXT_ID",
  },
  Object {
    "type": "CLEAR_CURRENT_TOOL_DATA",
  },
  Object {
    "type": "CLEAR_RESOURCES_REDUCER",
  },
  Object {
    "type": "CLEAR_PREVIOUS_FILTERS",
  },
  Object {
    "currentToolTitle": "",
    "type": "SET_CURRENT_TOOL_TITLE",
  },
  Object {
    "alertMessage": [TypeError: Cannot read property 'bookId' of undefined],
    "loading": undefined,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "showProjectValidationStepper": false,
    "type": "TOGGLE_PROJECT_VALIDATION_STEPPER",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "boolean": true,
    "type": "TOGGLE_HOME_VIEW",
  },
  Object {
    "type": "RESET_PROJECT_DETAIL",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_DATA",
  },
  Object {
    "type": "CLEAR_PREVIOUS_GROUPS_INDEX",
  },
  Object {
    "type": "CLEAR_CONTEXT_ID",
  },
  Object {
    "type": "CLEAR_CURRENT_TOOL_DATA",
  },
  Object {
    "type": "CLEAR_RESOURCES_REDUCER",
  },
  Object {
    "type": "CLEAR_PREVIOUS_FILTERS",
  },
  Object {
    "currentToolTitle": "",
    "type": "SET_CURRENT_TOOL_TITLE",
  },
  Object {
    "type": "CLEAR_COPYRIGHT_CHECK_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "type": "CLEAR_MERGE_CONFLICTS_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "projects": Array [
      Object {
        "accessTimeAgo": "in a few seconds",
        "bookAbbr": "tit",
        "bookName": "Titus",
        "isSelected": false,
        "projectName": "en_tit",
        "projectSaveLocation": "/Users/blm/translationCore/projects/en_tit",
        "target_language": Object {
          "direction": "ltr",
          "id": "en",
          "name": "English",
        },
      },
    ],
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "LOADED_ONLINE_FAILED",
  },
]
`;
